### PR TITLE
feat: add requiredClientCertificate on mutual TLS to make it optional

### PR DESCRIPTION
**Pull request by zufardhiyaulhaq**
**What type of PR is this?**
support https://github.com/envoyproxy/gateway/issues/3198
making client certificate validation optional

**What this PR does / why we need it**:
by making client certificate validation optional, we can add make 1 domain support both mutual TLS and TLS and we can filter APIs that require mTLS with https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#x-forwarded-client-cert later.

most API gateways support this mechanism, for example, Kong or Tyk.

    https://docs.konghq.com/hub/kong-inc/mtls-auth/
    https://tyk.io/docs/basic-config-and-security/security/mutual-tls/client-mtls/

**Which issue(s) this PR fixes**:
https://github.com/envoyproxy/gateway/issues/3198

### CODE REVIEW

**Comment by zufardhiyaulhaq** [comment]
@arkodg

File: api/v1alpha1/tls_types.go [hunk 1]
```diff
@@ -110,6 +110,12 @@ const (
 // to the Gateway.
 // By default, no client specific configuration is validated.
 type ClientValidationContext struct {
+	// RequiredClientCertificate set to true means that Envoy will reject connections without a valid client certificate.
```
Comment (arkodg) [inline api/v1alpha1/tls_types.go:113]: I'd flip the meaning of the noun here, so the default value is `false`
The two terms that come to mind are `optional` and `allowMissing`
https://github.com/envoyproxy/gateway/pull/3157 is trying to address a similar problem for JWT

**Comment by arkodg** [comment]
thanks for the quick PR, it looks good ! the only comment I have is around naming

File: api/v1alpha1/tls_types.go (hunk 1 above)
Comment (zufardhiyaulhaq) [reply api/v1alpha1/tls_types.go:113]: we cannot set the default value to false to support backward compatibility.
should we use `DisabledClientCertificateValidation`?

Comment (arkodg) [reply api/v1alpha1/tls_types.go:113]: thats too wordy imo

Comment (zufardhiyaulhaq) [reply api/v1alpha1/tls_types.go:113]: since it's under ClientValidation, should we set this as Enabled or Disabled? hmm

Comment (arkodg) [reply api/v1alpha1/tls_types.go:113]: the meaning of the noun picked, should convey the default meaning, e.g. `optional` or `allowMissing`  highlights that the default is `false`

Comment (zufardhiyaulhaq) [reply api/v1alpha1/tls_types.go:113]: if that is the case, I propose with `Required` and the default is true.
since it's under ClientValidationContext struct.

or do you have better name @arkodg?

Comment (arkodg) [reply api/v1alpha1/tls_types.go:113]: I prefer `optional` which defaults to `false`
will bring this up in the community meeting this week to speed up consensus on this

Comment (zufardhiyaulhaq) [reply api/v1alpha1/tls_types.go]: I am using `Required` and set default to `true`.
I am not sure why I cannot use `!` whn using optional. somehow translator_test.go failing

Comment (zhaohuabing) [reply api/v1alpha1/tls_types.go]: Both names sound good to me. However, I prefer to use `false` as default to keep consistent with other APIs.

Comment (zufardhiyaulhaq) [reply api/v1alpha1/tls_types.go]: Hi @arkodg @zhaohuabing change `Required` to `Optional` and use false by default.
please help recheck! thanks

File: api/v1alpha1/tls_types.go [hunk 2]
```diff
@@ -110,6 +110,10 @@ const (
 // to the Gateway.
 // By default, no client specific configuration is validated.
 type ClientValidationContext struct {
+	// Optional set to false means that Envoy will reject connections without a valid client certificate.
```
Comment (arkodg) [inline api/v1alpha1/tls_types.go:113]:
```suggestion
	// Optional set to true accepts connections even when a client doesn't present a certificate.
	// Defaults to false, which rejects connections without a valid client certificate.
```

File: internal/ir/xds.go [hunk 3]
```diff
@@ -281,6 +281,8 @@ type TLSConfig struct {
 	Certificates []TLSCertificate `json:"certificates,omitempty" yaml:"certificates,omitempty"`
 	// CACertificate to verify the client
 	CACertificate *TLSCACertificate `json:"caCertificate,omitempty" yaml:"caCertificate,omitempty"`
+	// RequiredClientCertificate to enforce client certificate
+	RequiredClientCertificate bool `json:"requiredClientCertificate,omitempty" yaml:"requiredClientCertificate,omitempty"`
```
Comment (arkodg) [inline internal/ir/xds.go:285]:
```suggestion
	RequireClientCertificate bool `json:"requireClientCertificate,omitempty" yaml:"requiredClientCertificate,omitempty"`
```

**Comment by arkodg** [comment]
minor comments around naming and doc string, else LGTM

**Comment by guydc** [comment]
@zufardhiyaulhaq - can you please rerun code-generation?

**Comment by zufardhiyaulhaq** [comment]
@guydc done, please check

**Review by guydc — APPROVED** [review]
LGTM. 

Would be great if we could also add an e2e test for this feature in a follow-up PR, but I guess that this is blocked on #2614 for now.

**Review by arkodg — APPROVED** [review]
LGTM thanks !
